src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,189 @@
+  scheduling::NodeID better_match_node;
+  int better_match_count = 0;
+  for (const auto &[node_id, count] : match_count_map) {
+    if (count == soft_expressions_size) {
+      all_match_nodes.push_back(node_id);
+    } else if (count > better_match_count) {
+      better_match_count = count;
+      better_match_node = node_id;
+    }
+  }
+  if (!all_match_nodes.empty()) {
+    std::uniform_int_distribution<int> distribution(0, all_match_nodes.size() - 1);
+    int node_index = distribution(gen_);
+    return all_match_nodes[node_index];
+  }
+  if (better_match_count > 0) {
+    return better_match_node;
+  }
+  return scheduling::NodeID::Nil();
+}
+
+void NodeLabelSchedulingPolicy::SelectNodesByHardExpressions(
+    absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const rpc::LabelMatchExpressions &expressions) const {
+  for (const auto &expression : expressions.expressions()) {
+    if (candidate_nodes.empty()) {
+      return;
+    }
+
+    for (auto iter = candidate_nodes.begin(); iter != candidate_nodes.end(); iter++) {
+      if (!IsNodeMatchLabelExpression(*(iter->second), expression)) {
+        candidate_nodes.erase(iter);
+      }
+    }
+  }
+}
+
+absl::flat_hash_map<scheduling::NodeID, int>
+NodeLabelSchedulingPolicy::ScoreNodesBySoftExpressions(
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const rpc::LabelMatchExpressions &expressions) const {
+  absl::flat_hash_map<scheduling::NodeID, int> match_count_map;
+  for (const auto &expression : expressions.expressions()) {
+    for (const auto &[node_id, node] : candidate_nodes) {
+      if (IsNodeMatchLabelExpression(*node, expression)) {
+        match_count_map[node_id]++;
+      }
+    }
+  }
+  return match_count_map;
+}
+
+bool NodeLabelSchedulingPolicy::IsNodeMatchLabelExpression(
+    const Node &node, const rpc::LabelMatchExpression &expression) const {
+  const auto &key = expression.key();
+  const auto &match_operator = expression.operator_();
+  if (match_operator.has_label_exists()) {
+    if (IsNodeLabelKeyExists(node, key)) {
+      return true;
+    }
+  } else if (match_operator.has_label_does_not_exist()) {
+    if (!IsNodeLabelKeyExists(node, key)) {
+      return true;
+    }
+  } else if (match_operator.has_label_in()) {
+    std::vector<std::string> values;
+    for (const auto &value : match_operator.label_in().values()) {
+      values.push_back(value);
+    }
+    if (IsNodeLabelInValues(node, key, values)) {
+      return true;
+    }
+  } else if (match_operator.has_label_not_in()) {
+    std::vector<std::string> values;
+    for (const auto &value : match_operator.label_not_in().values()) {
+      values.push_back(value);
+    }
+    if (!IsNodeLabelInValues(node, key, values)) {
+      return true;
+    }
+  } else {
+    RAY_CHECK(false) << "Node label match operator type must be one of `label_in`、"
+                        "`label_not_in`、`label_exists` or `label_does_not_exist`.";
```

Comment:
```suggestion
    RAY_CHECK(false) << "Node label match operator type must be one of `In`、"
                        "`NotIn`、`Exists` or `DoesNotExist`.";
```

Reply:
Since the MatchOperator name in the internal proto are "label_in"  "label_not_in," I chose to use those names.
![image](https://github.com/ray-project/ray/assets/11072802/eb3382c2-ae9a-4506-91bf-517d8bd6982f)